egui_glium/src/painter.rs

```diff
@@ -1,192 +1,386 @@
+
+        gl.viewport(0, 0, width_in_pixels as i32, height_in_pixels as i32);
+
+        gl.use_program(Some(self.program));
+
+        // The texture coordinates for text are so that both nearest and linear should work with the egui font texture.
+        // For user textures linear sampling is more likely to be the right choice.
+        gl.uniform_2_f32(Some(&self.u_screen_size), width_in_points, height_in_points);
+        gl.uniform_1_i32(Some(&self.u_sampler), 0);
+        gl.active_texture(glow::TEXTURE0);
+
+        gl.bind_vertex_array(Some(self.va));
+        gl.bind_buffer(glow::ARRAY_BUFFER, Some(self.vb));
+        gl.bind_buffer(glow::ELEMENT_ARRAY_BUFFER, Some(self.eb));
+        (width_in_pixels, height_in_pixels)
+    }
+
     /// Main entry-point for painting a frame.
     /// You should call `target.clear_color(..)` before
     /// and `target.finish()` after this.
-    pub fn paint_meshes<T: glium::Surface>(
+    ///
+    /// The following OpenGL features will be set:
+    /// - Scissor test will be enabled
+    /// - Cull face will be disabled
+    /// - Blend will be enabled
+    ///
+    /// The scissor area and blend parameters will be changed.
+    ///
+    /// As well as this, the following objects will be rebound:
+    /// - Vertex Array
+    /// - Vertex Buffer
+    /// - Element Buffer
+    /// - Texture (and active texture will be set to 0)
+    /// - Program
+    ///
+    /// Please be mindful of these effects when integrating into your program, and also be mindful
+    /// of the effects your program might have on this code. Look at the source if in doubt.
+    pub fn paint_meshes(
         &mut self,
-        display: &glium::Display,
-        target: &mut T,
+        gl_window: &glutin::WindowedContext<glutin::PossiblyCurrent>,
+        gl: &glow::Context,
         pixels_per_point: f32,
         cipped_meshes: Vec<egui::ClippedMesh>,
         egui_texture: &egui::Texture,
     ) {
-        self.upload_egui_texture(display, egui_texture);
-        self.upload_pending_user_textures(display);
+        #[cfg(debug_assertions)]
+        if self.destructed {
+            unreachable!();
+        }
 
+        self.upload_egui_texture(gl, egui_texture);
+        self.upload_pending_user_textures(gl);
+
+        let (w, h) = unsafe { self.prepare_painting(gl_window, gl, pixels_per_point) };
         for egui::ClippedMesh(clip_rect, mesh) in cipped_meshes {
-            self.paint_mesh(target, display, pixels_per_point, clip_rect, &mesh)
+            self.paint_mesh(gl, w, h, pixels_per_point, clip_rect, &mesh)
         }
+
+        assert_eq!(unsafe { gl.get_error() }, glow::NO_ERROR);
     }
 
     #[inline(never)] // Easier profiling
-    pub fn paint_mesh<T: glium::Surface>(
+    fn paint_mesh(
         &mut self,
-        target: &mut T,
-        display: &glium::Display,
+        gl: &glow::Context,
+        width_in_pixels: u32,
+        height_in_pixels: u32,
```

Comment:
```suggestion
        size_in_pixels: [u32; 2],
```

it's nice to clump together values that belong together